cmd/root_test.go

```diff
@@ -66,3 +67,117 @@ func TestRootCmd_Flags(t *testing.T) {
+      --data-dir string                                Data Directory for k0s \(default: .+k0s\)\. DO NOT CHANGE for an existing setup, things will break!
+  -d, --debug                                          Debug logging \(default: false\)
+      --debugListenOn string                           Http listenOn for Debug pprof handler \(default ":6060"\)
+      --disable-components strings                     disable components \(valid items: applier-manager,autopilot,control-api,coredns,csr-approver,endpoint-reconciler,helm,konnectivity-server,kube-controller-manager,kube-proxy,kube-scheduler,metrics-server,network-provider,node-role,system-rbac,windows-node,worker-config\)
+      --enable-cloud-provider                          Whether or not to enable cloud provider support in kubelet
+      --enable-dynamic-config                          enable cluster-wide dynamic config based on custom resource
+      --enable-k0s-cloud-provider                      enables the k0s-cloud-provider \(default false\)
+      --enable-metrics-scraper                         enable scraping metrics from the controller components \(kube-scheduler, kube-controller-manager\)
+      --enable-worker                                  enable worker \(default false\)
+  -h, --help                                           help for controller
+      --ignore-pre-flight-checks                       continue even if pre-flight checks fail
+      --iptables-mode string                           iptables mode \(valid values: nft, legacy, auto\)\. default: auto
+      --k0s-cloud-provider-port int                    the port that k0s-cloud-provider binds on \(default 10258\)
+      --k0s-cloud-provider-update-frequency duration   the frequency of k0s-cloud-provider node updates \(default 2m0s\)
+      --kube-controller-manager-extra-args string      extra args for kube-controller-manager
+      --kubelet-extra-args string                      extra args for kubelet
+      --labels strings                                 Node labels, list of key=value pairs
+  -l, --logging stringToString                         Logging Levels for the different components \(default \[.+]\)
+      --no-taints                                      disable default taints for controller node
+      --profile string                                 worker profile to use on the node \(default "default"\)
+      --single                                         enable single node \(implies --enable-worker, default false\)
+      --status-socket string                           Full file path to the socket file\. \(default: <rundir>/status\.sock\)
+      --taints strings                                 Node taints, list of key=value:effect strings
+      --token-file string                              Path to the file containing join-token\.
+  -v, --verbose                                        Verbose logging \(default: false\)
+$`, out.String())
```

Comment:
We have a bunch of flag sets that are picked up and remixed across subcommands, which makes it challenging to guess what changed to the args when changing the code. I thought it would be good to have the complete output in the tests, so that we actually notice if anything user-visible changes.

The reason it is a regex and not a simple string (which would be much nicer) is because of Linux/Windows differences in paths and spf13/pflag#362 (this affects the logging parameter). The former could be maybe tackled with fmt.Sprintf, the latter makes the whole output non-deterministic.

Creating this regex was a little but fiddly, but maintaining it will be much easier I suppose. K0s doesn't have much churn around its CLI args.

I'll have a look if I can come up with a way to fix the non-determinism. Then we could probably use plain strings here.

Reply:
It's a string now. PTAL.